components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc

```diff
@@ -0,0 +1,390 @@
+namespace brave_wallet {
+
+ZCashShieldSyncService::OrchardBlockScannerProxy::OrchardBlockScannerProxy(
+    std::array<uint8_t, kOrchardFullViewKeySize> full_view_key) {
+  background_block_scanner_.emplace(
+      base::ThreadPool::CreateSequencedTaskRunner({base::MayBlock()}),
+      full_view_key);
+}
+
+ZCashShieldSyncService::OrchardBlockScannerProxy::~OrchardBlockScannerProxy() =
+    default;
+
+void ZCashShieldSyncService::OrchardBlockScannerProxy::ScanBlocks(
+    std::vector<OrchardNote> known_notes,
+    std::vector<zcash::mojom::CompactBlockPtr> blocks,
+    base::OnceCallback<void(base::expected<OrchardBlockScanner::Result,
+                                           OrchardBlockScanner::ErrorCode>)>
+        callback) {
+  background_block_scanner_.AsyncCall(&OrchardBlockScanner::ScanBlocks)
+      .WithArgs(std::move(known_notes), std::move(blocks))
+      .Then(std::move(callback));
+}
+
+ZCashShieldSyncService::ZCashShieldSyncService(
+    ZCashRpc* zcash_rpc,
+    const mojom::AccountIdPtr& account_id,
+    const mojom::ZCashAccountShieldBirthdayPtr& account_birthday,
+    const std::array<uint8_t, kOrchardFullViewKeySize>& fvk,
+    base::FilePath db_dir_path)
+    : zcash_rpc_(zcash_rpc),
+      account_id_(account_id.Clone()),
+      account_birthday_(account_birthday.Clone()),
+      db_dir_path_(db_dir_path) {
+  chain_id_ = GetNetworkForZCashKeyring(account_id->keyring_id);
+  block_scanner_ = std::make_unique<OrchardBlockScannerProxy>(fvk);
+  background_orchard_storage_.emplace(
+      base::ThreadPool::CreateSequencedTaskRunner({base::MayBlock()}),
+      db_dir_path_);
+}
+
+ZCashShieldSyncService::~ZCashShieldSyncService() = default;
+
+void ZCashShieldSyncService::SetOrchardBlockScannerProxyForTesting(
+    std::unique_ptr<OrchardBlockScannerProxy> block_scanner) {
+  block_scanner_ = std::move(block_scanner);
+}
+
+void ZCashShieldSyncService::StartSyncing(
+    mojo::PendingRemote<mojom::ZCashSyncObserver> observer) {
+  if (observer_.is_bound()) {
+    NOTREACHED_IN_MIGRATION();
+    return;
+  }
+  observer_.Bind(std::move(observer));
+  ScheduleWorkOnTask();
+}
+
+void ZCashShieldSyncService::PauseSyncing() {
+  stopped_ = true;
+  observer_->OnStop();
+}
+
+mojom::ZCashShieldSyncStatusPtr ZCashShieldSyncService::GetSyncStatus() {
+  return current_sync_status_.Clone();
+}
+
+void ZCashShieldSyncService::ScheduleWorkOnTask() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE, base::BindOnce(&ZCashShieldSyncService::WorkOnTask,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::WorkOnTask() {
+  if (stopped_) {
+    return;
+  }
+
+  if (error_) {
+    if (observer_) {
+      observer_->OnError(base::NumberToString(error_->code) + ": " +
+                         error_->message);
+    }
+    return;
+  }
+
+  if (!chain_tip_block_) {
+    UpdateChainTip();
+    return;
+  }
+
+  if (!latest_scanned_block_) {
+    GetAccountMeta();
```

Comment:
Can we split chain of callbacks behind GetAccountMeta into more ifs in ::WorkOnTask ?

Reply:
Done